### PR TITLE
pkg/package_archive

### PR DESCRIPTION
**Pull request by isoos**
- #2227
- I wanted to have a design where we are able to display all of the issues at once for the user before uploader, but also to have an easy way to throw an exception on it (as before) to keep it the same way on the server as before.

### CODE REVIEW

File: pkg/package_archive/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: package_archive
+publish_to: none
+
```
Comment (jonasfj) [inline pkg/package_archive/pubspec.yaml]: Please give this a description and maybe a `README` that just says what this package does..

File: pkg/package_archive/lib/src/tar_utils.dart [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+/// Reads a text content of [name] from the tar.gz file identified by [path].
+///
+/// When [maxLength] is specified, the text content is cut at [maxLength]
+/// characters (with `[...]\n\n` added to it).
+Future<String> readTarballFile(String path, String name,
+    {int maxLength = 0}) async {
+  final result = await Process.run(
+    'tar',
+    ['-O', '-xzf', path, name],
```
Comment (jonasfj) [inline pkg/package_archive/lib/src/tar_utils.dart]: What happens if this a 10GB sparse file?

File: pkg/package_archive/lib/package_archive.dart [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+import 'src/names.dart';
+import 'src/tar_utils.dart';
+
+// The maximum stored length of `README.md` and other user-provided file content
+// that is stored separately in the database.
+final _maxStoredLength = 128 * 1024;
+
+/// A validation error in the package archive.
+class ArchiveError {
```
Comment (jonasfj) [inline pkg/package_archive/lib/package_archive.dart]: Maybe call it an `Issue` as `Error` indicates subclassing `Error`.

File: pkg/package_archive/lib/package_archive.dart [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  final String message;
+
+  ArchiveError(this.message);
+
+  @override
+  String toString() => message;
+}
+
+/// The observed / extracted information of a package archive.
+class PackageArchive {
```
Comment (jonasfj) [inline pkg/package_archive/lib/package_archive.dart]: Is it perhaps a `PackageSummary` instead?

File: pkg/package_archive/lib/package_archive.dart [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    this.examplePath,
+    this.exampleContent,
+    this.libraries,
+  });
+
+  bool get hasError => errors != null && errors.isNotEmpty;
+}
+
+/// Observe the .tar.gz archive on [archivePath] and return the results.
+Future<PackageArchive> observePackageArchive(String archivePath) async {
```
Comment (jonasfj) [inline pkg/package_archive/lib/package_archive.dart]: Could call it `summarizePackage`?

File: pkg/package_archive/README.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# package_archive
```
Comment (jonasfj) [inline pkg/package_archive/README.md]: Couldn't it be called `package_reader` or something else? `pub_package_reader` or `package_validator`?

File: pkg/package_archive/lib/package_archive.dart [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  final libraries = files
+      .where((file) => file.startsWith('lib/'))
+      .where((file) => !file.startsWith('lib/src'))
+      .where((file) => file.endsWith('.dart'))
+      .map((file) => file.substring('lib/'.length))
+      .toList();
+
+  errors.addAll(validatePackageName(pubspec.name));
+  errors.addAll(syntaxCheckHomepageUrl(
+      pubspec.homepage ?? pubspec.repository?.toString()));
```
Comment (jonasfj) [inline pkg/package_archive/lib/package_archive.dart]: By increasing the level of abstraction I mean... we probably need something like:
https://github.com/dart-lang/pub/tree/master/lib/src/validator

So, maybe we have a `Package` class which has a function to list and read files... and long with parsed pubspec, and README, etc...

We would then have a long list of `Validators` that are passed a `Package` instance, and then return a list of `Issues`.

This way, each `Validator` is probably just a function... we can have them in separate files, and we can easily switch on/off which ones are active by adding removing the functions to/from a list of functions.

Just as importantly, these validators can be tested independently, and each one of them would deal with a separate and independent concern. One for `syntaxCheckHomepageUrl` another for `validatePackageName`.... maybe a different one to `pingHomepageUrl`, etc...

File: pkg/package_archive/lib/src/tar_utils.dart (hunk 2 above)
Comment (isoos) [reply pkg/package_archive/lib/src/tar_utils.dart]: I'd treat this PR as a refactoring, and don't add further checks. We have a check at upload time to block upload if it is larger than a certain size, but it wouldn't hurt to double-check here.

File: pkg/package_archive/lib/package_archive.dart (hunk 7 above)
Comment (isoos) [reply pkg/package_archive/lib/package_archive.dart]: Agreed, but let's do it in a separate PR.

**Pull request merged**